Add functionality to specify private network

iamsounoringinal opened the pull request:
##### SUMMARY
Adds the 'network_id' parameter to allow specification of a private network id. This handles the case where a Vultr account has multiple private networks.

This takes into account Vultr's API requirement that only 'NETWORKID' is set or 'enable_private_network', not both.

Does not change current behaviour.

Documentation section updated to reflect change.

##### ISSUE TYPE

 - Feature Pull Request
##### COMPONENT NAME
vultr_server module

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/Users/hhynd/.ansible/library']
  ansible python module location = /Library/Python/2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.10 (default, Oct  6 2017, 22:29:07) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```


##### ADDITIONAL INFORMATION
The current behaviour doesn't allow you to select a private network to join on server creation. 

If there are more than one private network configured in the Vultr region, Vultr will return as below: 
```
fatal: [localhost -> localhost]: FAILED! => {"changed": true, "msg": "URL https://api.vultr.com/v1/server/create, method POST with data VPSPLANID=201&hostname=[FQDN REMOVED]]&DCID=19&user_data=[API-KEY REMOVED]&notify_activate=no&label=server-build-test2&enable_private_network=yes&SCRIPTID=163879&tag=corporate&FIREWALLGROUPID=96484181&OSID=215. Returned 412, with body: HTTP Error 412: Request Failed Server add failed: No NETWORKID defined, but multiple private networks exist in this location.", "vultr_api": {"api_account": "default", "api_endpoint": "https://api.vultr.com", "api_retries": 5, "api_timeout": 60}, "vultr_server": {}}
```

This change allows NETWORKID specification and honours Vultr's API specification. ie. enable_private_network can be specificied, or NETWORKID, not both.

hhynd committed:
Add functionality to specify private network

Adds the 'network_id' parameter to allow specification of a private network id.

This takes into account Vultr's API requirement that only 'NETWORKID' is set or 'enable_private_network', not both.

Does not change current behaviour.

Documentation section updated to reflect change.

hhynd committed:
Remove 'enable_private_network' from data object

Fixing an oversite.

Should've removed this to ensure correct functionality.

hhynd committed:
Adding version_added to documentation

Spredzy commented on lib/ansible/modules/cloud/vultr/vultr_server.py:
```diff
@@ -54,6 +54,12 @@
     description:
       - Whether to enable private networking or not.
     type: bool
+  network_id:
```
Based on the documentation https://www.vultr.com/api/#server_create

> NETWORKID array (optional) List of private networks to attach to this server.  Use either this field or enable_private_network, not both

So I would make it a lists. Also for naming/user experience, I'd rename it networks where user can input a lists of networks name (easier to read/understand). Later a `get_networks()` method could be used to translate from name to id to pass values in the payload.

resmo replied:
I am also +1 what @Spredzy suggested.

Spredzy reviewed (changes requested):
Thanks @iamsounoringinal for this contribution, it is much appreciated. I've left a small comment regarding the interface this module offers. Plus you should add some tests with this pull request. Tests are available in `test/legacy/vultr.yml`, if you have any question regarding how it works, feel free to ask  any question.

hhynd committed:
Addresses suggested improvements

Module Changes:
- Documentation changed to reflect changed behaviour
- Example added for usage
- method to get list of private networks from Vultr API added based on a list of network names
- parameter name change from network_id to networks, and type from string to list

Module-util Changes:
-  adds method "query_resources_by_key" to support multiple values being returned by the API. Maintains current functionality of "query_resource_by_key" so as to avoid too many changes to otherwise functional areas of the module.

hhynd committed:
Updating "version added"

iamsounoringinal commented:
Have addressed feature behviour changes, but not tests. which have so far been a bit beyond my ability.

hhynd committed:
adress pep8 test failure

hhynd committed:
Address suggestions from Resmo

- Modifies the documentation slightly
- adds "network" as an alias for "networks" to allow a single network to be supplied in a more appropriate way
- adds mutually exclusive prpoperty

iamsounoringinal commented:
Thanks @resmo, those were great suggestions :-)

iamsounoringinal commented:
Whats the next step here guys?

resmo commented:
The way `query_resources_by_key` is implemented is not very resource friendly, however I think we could go this way for now and improve later. @Spredzy your opinion?

Merge conflicts must be addressed of course, before we could merge.
